### ISSUE TITLE
Request - Apprise API?

### ISSUE THREAD

**Issue by nebb00** [engels74/mover-status#5]
Just wondering if you would be interested adding Apprise support. Its just an curl using the right formatting, that way any notification service could be used. Just an idea.

https://github.com/caronc/apprise-api

**Comment by engels74** [comment]
I'll look into it, no promises :).

But I'm thinking of restructuring this little project a bit, or doing an off-shoot version of it.. Probably in Python or something, where I can split up the project to multiple files. I don't think currently the user scripts plugin in Unraid supports this properly.

Either way, it's going to be a while before I'll take a proper look at this project again. I'll happily accept a PR, if anyone's up for it.

Reason being the bash script is getting increasingly complex, and that combined with me not being that great of a programmer, makes it a bit hard to manage. Especially if I keep on adding more and more notification services, that all has different approaches to how they accept/decode e.g. markdown or formatting in general.

So my plan is to make it available as a Docker image on the Unraid, that then uses pid=host or uses the `/proc` in read-only mode or something, if that works.

**Comment by RyanRoberts210** [comment]
I just wanted to second this great idea!  It is so surprising that this functionality (i.e. mover status reporting) isn't just built in to Unraid itself...  In addition to apprise there is always the built-in notification processes in Unraid itself which could be used.  I totally get the length of the script but have to say it is immaculate in layout and documentation, I can follow all of it.  :)  I would love to have this functionality but don't have / use the services supported...  Thanks for your consideration!

**Comment by engels74** [comment]
> I just wanted to second this great idea! It is so surprising that this functionality (i.e. mover status reporting) isn't just built in to Unraid itself... In addition to apprise there is always the built-in notification processes in Unraid itself which could be used. I totally get the length of the script but have to say it is immaculate in layout and documentation, I can follow all of it. :) I would love to have this functionality but don't have / use the services supported... Thanks for your consideration!

I know Unraid includes a notification system, but it's quite basic. Personally, I prefer the way the bash script is handling it currently, particularly the Discord one, because it is more visually appealing. That required a significant amount of effort. I tried Unraid's notification system a while ago, and it felt a little lackluster. Or maybe I don't know how to use it properly :).

When I have the time and motivation, I'll consider undertaking a restructuring or creating it from scratch in Python. To utilize a simple Docker image in Unraid. This should make it easy for me to add multiple/several new notification services, such as the Apprise API, Gotify and ntfy.

But I don't have the motivation (or time) right now, especially since the script is currently working well for me. So, once again, if anyone wants to make any PRs with a new notification service, I'll gladly merge it in. Especially considering it'll probably be a while before I get to complete a rework of the project.

**Comment by btc08gh** [comment]
I implemented Apprise here if you want to test it out:

https://github.com/btc08gh/mover-status/tree/feature/unraid-notifications

(It also includes using the Apprise CLI as well as local Unraid system notifications.)

@engels74 I will create a PR for this if you’d like.  Just let me know your wishes.

**Comment by engels74** [comment]
> I implemented Apprise here if you want to test it out:
> 
> https://github.com/btc08gh/mover-status/tree/feature/unraid-notifications
> 
> (It also includes using the Apprise CLI as well as local Unraid system notifications.)
> 
> [@engels74](https://github.com/engels74) I will create a PR for this if you’d like. Just let me know your wishes.

Yeah, at this point I should probably merge the other one you made. Am still planning to do a proper Python rewrite of this, with suitable plugin support, but that'll probably be a few months out.

As long as you have tested things sufficiently, you can create the PR :).

**Comment by btc08gh** [comment]
I deliberately stacked the work to make the eventual merges easier to manage.

The progression in my fork is essentially:

Pushover → Apprise → native Unraid notifications

I kept each feature in separate commits/branches so once the Pushover PR is merged, I can cut a fresh Apprise branch from your updated main and cherry-pick only the Apprise commits. That way the Apprise PR you review will contain only the Apprise changes, rather than dragging the already reviewed Pushover diff along with it. I’m planning to do the same thing with a native Unraid notification support afterward.

I’ve also tested the Apprise work pretty heavily on my actual Unraid server. The testing so far includes:

* Apprise CLI mode using the Lime Technology Apprise package (/usr/bin/apprise)
* Successful real Pushover delivery through Apprise CLI
* Rejected/broken targets returning failure correctly
* Mover Status DRY_RUN success and failure paths through CLI
* Multiple Apprise targets with per-target retry isolation and a failed target retries without duplicating notifications to healthy targets
* Progress retry/coalescing so failed targets receive the newest scheduled progress rather than stale updates
* Completion retry isolation
* Full synthetic Mover lifecycle through CLI: initial/progress updates through 25/50/75/99% and completion
* Apprise API mode using the LinuxServer Apprise API container
* Successful and failed stateless /notify calls
* Mover Status DRY_RUN through the API, including successful and rejected targets
* Multiple-target progress and completion isolation through the API
* API outage/recovery testing where I stopped the API, confirmed the notification remained pending, restarted it and verified successful retry
* HTTP response handling for successful 2xx, rejected 424, and an intentional 302
* Full synthetic Mover lifecycle through API mode
* CLI regression testing after adding API support
* Invalid APPRISE_MODE and invalid API URL validation
* Real-world API testing on my Unraid server through my HTTPS reverse proxy, including an actual Pushover notification delivered through the remote Apprise API
* bash -n and git diff --check throughout

I also documented the two Unraid packages I actually tested with rather than just assuming compatibility.

So yes, I’m comfortable with the Apprise implementation from a testing standpoint. Once the Pushover PR is merged, I’ll rebuild the Apprise branch cleanly from the new main, verify the resulting diff contains only Apprise, rerun the final checks, and send you the PR.

And the Python rewrite sounds great!

**Comment by engels74** [comment]
Sounds great 👍 . 

Whenever you have time, open up the Apprise PR, and I'll merge it and create/bump a new release of `moverStatus.sh`

**Comment by engels74** [comment]
Closed with the merge of https://github.com/engels74/mover-status/pull/28
